File: Code Samples/Stack - Nikita Choudhari/Python/Stack.py
class CustomStack:
    def __init__(self, size):
        self.max_size = size
        self.stack_array = [None] * size
        self.top = -1

    def push(self, item):
        if self.is_full():
            print(f"Stack is full. Cannot push item: {item}")
            return
        self.top += 1
        self.stack_array[self.top] = item

    def pop(self):
        if self.is_empty():
            print("Stack is empty. Cannot pop an item.")
            return None
        popped_item = self.stack_array[self.top]
        self.top -= 1
        return popped_item

    def peek(self):
        if self.is_empty():
            print("Stack is empty. Cannot peek.")
            return None
        return self.stack_array[self.top]

    def is_empty(self):
        return self.top == -1

    def is_full(self):
        return self.top == self.max_size - 1

    def size(self):
        return self.top + 1

stack = CustomStack(5)

stack.push(1)
stack.push(2)
stack.push(3)

print("Popped:", stack.pop())
print("Peek:", stack.peek())
print("Is empty?", stack.is_empty())

File: Code Samples/Stack - Nikita Choudhari/CPP/Stack.cpp
#include <iostream>

template <typename T>
class CustomStack {
private:
    int maxSize;
    T* stackArray;
    int top;

public:
    CustomStack(int size) : maxSize(size), top(-1) {
        stackArray = new T[size];
    }

    void push(T item) {
        if (isFull()) {
            std::cout << "Stack is full. Cannot push item: " << item << std::endl;
            return;
        }
        stackArray[++top] = item;
    }

    T pop() {
        if (isEmpty()) {
            std::cout << "Stack is empty. Cannot pop an item." << std::endl;
            return T();
        }
        return stackArray[top--];
    }

    T peek() {
        if (isEmpty()) {
            std::cout << "Stack is empty. Cannot peek." << std::endl;
            return T();
        }
        return stackArray[top];
    }

    bool isEmpty() {
        return top == -1;
    }

    bool isFull() {
        return top == maxSize - 1;
    }

    int size() {
        return top + 1;
    }

    ~CustomStack() {
        delete[] stackArray;
    }
};

int main() {
    CustomStack<int> stack(5);

    stack.push(1);
    stack.push(2);
    stack.push(3);

    std::cout << "Popped: " << stack.pop() << std::endl;
    std::cout << "Peek: " << stack.peek() << std::endl;
    std::cout << "Is empty? " << stack.isEmpty() << std::endl;

    return 0;
}
